File: internal-admin/frontend/lib/auth/session.ts
// The signed session cookie.
//
// Deliberately NOT "server-only" and deliberately free of node:crypto: this
// runs in middleware too, which is the Edge runtime. Web Crypto's HMAC exists
// in both, so one implementation covers both and there is only one place where
// a session is judged valid.
//
// The cookie carries the claims and a signature over them. It is signed, not
// encrypted — so never put anything in here that the holder shouldn't read.
// Tampering is what the signature prevents.

export interface Session {
  /** Lower-cased email. */
  email: string;
  name: string;
  role: "admin" | "viewer";
  /** True for the environment-configured account that always has access. */
  root: boolean;
  /** Seconds since the epoch. */
  exp: number;
}

export const SESSION_COOKIE = "mf_admin";
export const SESSION_MAX_AGE = 60 * 60 * 12; // 12 hours

const enc = new TextEncoder();

const b64url = (bytes: Uint8Array) => {
  let s = "";
  for (const b of bytes) s += String.fromCharCode(b);
  return btoa(s).replace(/\+/g, "-").replace(/\//g, "_").replace(/=+$/, "");
};

// Backed by a plain ArrayBuffer, which is what Web Crypto's BufferSource wants
// — Uint8Array.from can be typed over SharedArrayBuffer and won't assign.
const unb64url = (s: string): Uint8Array<ArrayBuffer> => {
  const pad = s.replace(/-/g, "+").replace(/_/g, "/") + "===".slice((s.length + 3) % 4);
  const bin = atob(pad);
  const view = new Uint8Array(new ArrayBuffer(bin.length));
  for (let i = 0; i < bin.length; i++) view[i] = bin.charCodeAt(i);
  return view;
};

async function key(secret: string): Promise<CryptoKey> {
  return crypto.subtle.importKey("raw", enc.encode(secret), { name: "HMAC", hash: "SHA-256" }, false, ["sign", "verify"]);
}

/**
 * The signing secret. Absent, every session is invalid and nobody can sign in
 * — which is the safe direction to fail. It is never defaulted to a constant,
 * because a known secret means anyone can mint an admin cookie.
 */
export function sessionSecret(): string | null {
  const s = process.env.SESSION_SECRET;
  return s && s.length >= 32 ? s : null;
}

export async function signSession(claims: Omit<Session, "exp">, secret: string, maxAge = SESSION_MAX_AGE): Promise<string> {
  const payload: Session = { ...claims, exp: Math.floor(Date.now() / 1000) + maxAge };
  const body = b64url(enc.encode(JSON.stringify(payload)));
  const sig = new Uint8Array(await crypto.subtle.sign("HMAC", await key(secret), enc.encode(body)));
  return `${body}.${b64url(sig)}`;
}

/** The claims if the signature holds and the session hasn't expired, else null. */
export async function readSession(token: string | undefined, secret: string | null): Promise<Session | null> {
  if (!token || !secret) return null;
  const dot = token.lastIndexOf(".");
  if (dot < 1) return null;
  const body = token.slice(0, dot);
  try {
    const ok = await crypto.subtle.verify("HMAC", await key(secret), unb64url(token.slice(dot + 1)), enc.encode(body));
    if (!ok) return null;
    const claims = JSON.parse(new TextDecoder().decode(unb64url(body))) as Session;
    if (typeof claims.exp !== "number" || claims.exp * 1000 < Date.now()) return null;
    if (typeof claims.email !== "string" || !claims.email) return null;
    return claims;
  } catch {
    return null;
  }
}

File: candidate/backend/internal/session/session.go
// Package session verifies the signed cookies the two Next.js frontends
// issue — it deliberately does not mint a first sign-in for either portal;
// sign-in stays where the password hash and scrypt work already live
// (candidate/frontend/src/lib/auth, internal-admin/frontend/lib/auth). This
// package is the Go-side twin of both apps' lib/auth/session.ts: same
// HMAC-SHA256-over-base64url construction, so a cookie either app signs
// verifies here without either app knowing the Go backend exists.
//
// The cookie is signed, not encrypted — the signature only proves the claims
// weren't tampered with after signing. Never rely on it to hide anything a
// holder of their own cookie shouldn't be able to read.
package session

import (
	"crypto/hmac"
	"crypto/sha256"
	"encoding/base64"
	"encoding/json"
	"errors"
	"strings"
	"time"
)

// Cookie names — must match SESSION_COOKIE in each frontend's session.ts.
const (
	CandidateCookie = "mf_candidate"
	AdminCookie     = "mf_admin"
)

// ErrInvalid covers every way a token can fail to verify: missing, malformed,
// wrong signature, or expired. Deliberately one error rather than several —
// none of those cases should be told apart by a caller (that's exactly the
// kind of detail an attacker probing the endpoint shouldn't get back).
var ErrInvalid = errors.New("session: invalid or expired token")

// CandidateClaims mirrors candidate/frontend/src/lib/auth/session.ts's
// exported Session interface field for field.
type CandidateClaims struct {
	ID    string `json:"id"`
	Email string `json:"email"`
	Name  string `json:"name"`
	Exp   int64  `json:"exp"`
}

// AdminClaims mirrors internal-admin/frontend/lib/auth/session.ts's Session.
type AdminClaims struct {
	Email string `json:"email"`
	Name  string `json:"name"`
	Role  string `json:"role"` // "admin" | "viewer"
	Root  bool   `json:"root"`
	Exp   int64  `json:"exp"`
}

func b64urlEncode(b []byte) string { return base64.RawURLEncoding.EncodeToString(b) }

func b64urlDecode(s string) ([]byte, error) { return base64.RawURLEncoding.DecodeString(s) }

// sign produces "<b64url(json)>.<b64url(hmac-sha256)>", matching the TS
// signSession's construction exactly: the signature covers the bytes of the
// base64url *string*, not the raw JSON bytes underneath it.
func sign(payload any, secret string) (string, error) {
	raw, err := json.Marshal(payload)
	if err != nil {
		return "", err
	}
	body := b64urlEncode(raw)
	mac := hmac.New(sha256.New, []byte(secret))
	mac.Write([]byte(body))
	return body + "." + b64urlEncode(mac.Sum(nil)), nil
}

// SignCandidate is used by tests and by any future Go-side session issuance;
// today candidate/frontend is the only real minter, this just proves the two
// implementations agree bit for bit.
func SignCandidate(c CandidateClaims, secret string) (string, error) { return sign(c, secret) }

// SignAdmin is the admin equivalent of SignCandidate.
func SignAdmin(a AdminClaims, secret string) (string, error) { return sign(a, secret) }

// verify checks the signature and expiry, returning the raw claims JSON for
// the caller to decode into whichever shape it expects. Constant-time
// comparison (hmac.Equal) throughout — this is the one function every
// authenticated request in the service runs through.
func verify(token, secret string) (json.RawMessage, error) {
	if token == "" || secret == "" {
		return nil, ErrInvalid
	}
	dot := strings.LastIndexByte(token, '.')
	if dot < 1 {
		return nil, ErrInvalid
	}
	body, sigPart := token[:dot], token[dot+1:]

	sig, err := b64urlDecode(sigPart)
	if err != nil {
		return nil, ErrInvalid
	}
	mac := hmac.New(sha256.New, []byte(secret))
	mac.Write([]byte(body))
	if !hmac.Equal(mac.Sum(nil), sig) {
		return nil, ErrInvalid
	}

	raw, err := b64urlDecode(body)
	if err != nil {
		return nil, ErrInvalid
	}

	var exp struct {
		Exp int64 `json:"exp"`
	}
	if err := json.Unmarshal(raw, &exp); err != nil {
		return nil, ErrInvalid
	}
	if exp.Exp <= 0 || time.Unix(exp.Exp, 0).Before(time.Now()) {
		return nil, ErrInvalid
	}

	return json.RawMessage(raw), nil
}

// VerifyCandidate validates an "mf_candidate" cookie value against
// CANDIDATE_SESSION_SECRET (the same value candidate/frontend's
// SESSION_SECRET holds).
func VerifyCandidate(token, secret string) (*CandidateClaims, error) {
	raw, err := verify(token, secret)
	if err != nil {
		return nil, err
	}
	var c CandidateClaims
	if err := json.Unmarshal(raw, &c); err != nil || c.ID == "" || c.Email == "" {
		return nil, ErrInvalid
	}
	return &c, nil
}

// VerifyAdmin validates an "mf_admin" cookie value against
// ADMIN_SESSION_SECRET (the same value internal-admin/frontend's
// SESSION_SECRET holds).
func VerifyAdmin(token, secret string) (*AdminClaims, error) {
	raw, err := verify(token, secret)
	if err != nil {
		return nil, err
	}
	var a AdminClaims
	if err := json.Unmarshal(raw, &a); err != nil || a.Email == "" || (a.Role != "admin" && a.Role != "viewer") {
		return nil, ErrInvalid
	}
	return &a, nil
}
